server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraDumbBlobStoreCache.java

```diff
@@ -0,0 +1,128 @@
+import static org.apache.james.blob.cassandra.BlobTables.DumbBlobCache.TABLE_NAME;
+import static org.apache.james.blob.cassandra.BlobTables.DumbBlobCache.TTL_FOR_ROW;
+
+import java.nio.ByteBuffer;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.blob.api.BlobId;
+import org.reactivestreams.Publisher;
+
+import com.datastax.driver.core.ConsistencyLevel;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Session;
+
+import reactor.core.publisher.Mono;
+
+public class CassandraDumbBlobStoreCache implements DumbBlobStoreCache {
+
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement insertStatement;
+    private final PreparedStatement selectStatement;
+    private final PreparedStatement deleteStatement;
+
+    private final int readTimeOutFromDataBase;
+    private final int timeToLive;
+
+    @Inject
+    public CassandraDumbBlobStoreCache(Session session, CassandraCacheConfiguration cacheConfiguration) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.insertStatement = prepareInsert(session);
+        this.selectStatement = prepareSelect(session);
+        this.deleteStatement = prepareDelete(session);
+
+        this.readTimeOutFromDataBase = (int) cacheConfiguration.getReadTimeOut().toMillis();
+        this.timeToLive = (int) cacheConfiguration.getTtl().getSeconds();
+    }
+
+    @Override
+    public Publisher<Void> cache(BlobId blobId, byte[] bytes) {
+        return Mono.just(bytes)
+            .flatMap(data -> Mono.from(save(blobId, ByteBuffer.wrap(data, 0, data.length))));
+    }
+
+    private Publisher<Void> save(BlobId blobId, ByteBuffer data) {
+        return cassandraAsyncExecutor.executeVoid(
+            insertStatement.bind()
+                .setString(ID, blobId.asString())
+                .setBytes(DATA, data)
+                .setInt(TTL_FOR_ROW, timeToLive)
+                .setConsistencyLevel(ConsistencyLevel.ONE));
+    }
+
+    @Override
+    public Publisher<byte[]> read(BlobId blobId) {
+        return cassandraAsyncExecutor
+            .executeSingleRow(
+                selectStatement.bind()
+                    .setString(ID, blobId.asString())
+                    .setConsistencyLevel(ConsistencyLevel.ONE)
+                    .setReadTimeoutMillis(readTimeOutFromDataBase)
+            )
+            .map(row -> {
+                ByteBuffer byteBuffer = row.getBytes(DATA);
+                byte[] data = new byte[byteBuffer.remaining()];
+                byteBuffer.get(data);
+                return data;
+            });
+    }
+
+    @Override
+    public Publisher<Void> remove(BlobId blobId) {
+        return cassandraAsyncExecutor.executeVoid(
+            deleteStatement.bind().setString(ID, blobId.asString()));
+    }
+
+    private PreparedStatement prepareDelete(Session session) {
```

Comment:
If I'm not wrong, our cache has a problem: deletion being eventually consistent, we can retrieve blobs from cache that should already be deleted from cached blobstore.

Reply:
Can i ask you how it happend?

Reply:
> If I'm not wrong, our cache has a problem: deletion being eventually consistent, we can retrieve blobs from cache that should already be deleted from cached blobstore.

I don't understand which elements makes you say so. A delete position a tumbstone in the memtable and in the commitlog. It is immediately applied by Cassandra.

Won't put Lightweight transaction and strong consistency in there.

We have good quality tests at the -dumb- blob store level to protect us about about such concurency corner cases

Reply:
1. save blob A
1.1. save blob A in blobstore
1.2. save blob A in cache
2. delete blob B
2.1. remove blob A from cache
2.2. remove blob A from blobstore
3. read blob A
3.1 read in cache actually retrieves the result because of CL ONE on delete/read requests

Am I wrong?

Reply:
Wrong. As RF will be 1, CL ONE will guaranty clearing the cache.

If that's still an issue for you we **might** increase the CL (to ALL ???) to unlock the ticket but I truely believe we don't have any issue here.

Reply:
Oh well, that's ok then, we don't have the bug. And yes, upgrading to ALL would make sense as it would prevent problems in case somebody tries to change the RF of this table.

Reply:
But it might endanger blob deletion availability. (nothing is free)

Reply:
isn't it already the case as long as RF = 1?

Reply:
OK to set consistency level to ALL upon deletes then.